src/VisualStudio/Core/Def/Implementation/TableDataSource/VisualStudioDiagnosticListTable.BuildTableDataSource.cs

```diff
@@ -238,6 +238,25 @@ public override bool TryNavigateTo(int index, NavigationOptions options, Cancell
                             return documentId;
                         }
 
+                        if (solution.GetProject(projectId) is { } project)
+                        {
+                            // We couldn't find a document matching a known ID when the item was created, so it may be a
+                            // source generator output.
+                            var documents = threadingContext.JoinableTaskFactory.Run(() => project.GetSourceGeneratedDocumentsAsync(cancellationToken).AsTask());
+                            if (documentId is not null)
```

Comment:
So do we have both cases here (document is null and not null) for DiagnosticTableItem reported on a source generated document? If so, do we know which diagnostic source is leading to each of these cases? IMO, this code feels like a hack and we should always have a non-null documentId for DiagnosticTableItem on a source generated document. This would be the case if we move all the code that is mapping diagnostic's reported file path to document ID to the diagnostic source that is creating the original DiagnosticData.